synapse/logging/opentracing.py

```diff
@@ -798,33 +797,31 @@ def extract_text_map(carrier: Dict[str, str]) -> Optional["opentracing.SpanConte
 # Tracing decorators
 
 
-def trace(func=None, opname: Optional[str] = None):
+def trace_with_opname(opname: str) -> Callable[[Callable[P, R]], Callable[P, R]]:
     """
-    Decorator to trace a function.
-    Sets the operation name to that of the function's or that given
-    as operation_name. See the module's doc string for usage
-    examples.
+    Decorator to trace a function with a custom opname.
+
+    See the module's doc string for usage examples.
+
     """
 
-    def decorator(func):
+    def decorator(func: Callable[P, R]) -> Callable[P, R]:
         if opentracing is None:
             return func  # type: ignore[unreachable]
 
-        _opname = opname if opname else func.__name__
-
         if inspect.iscoroutinefunction(func):
 
             @wraps(func)
-            async def _trace_inner(*args, **kwargs):
-                with start_active_span(_opname):
-                    return await func(*args, **kwargs)
+            async def _trace_inner(*args: P.args, **kwargs: P.kwargs) -> R:
+                with start_active_span(opname):
+                    return await func(*args, **kwargs)  # type: ignore[misc]
```

Comment:
For completeness this is:

```
synapse/logging/opentracing.py:819: error: Incompatible types in "await" (actual type "R", expected type "Awaitable[Any]")  [misc]
```

Reply:
Thanks. I think this is fine: within this branch, `func` is a coroutine so should return an awaitable. It's a shame mypy can't deduce this for itself, given [the use](https://github.com/python/typeshed/blob/6d0e01e62bb3236d387189509169b6ac40012567/stdlib/inspect.pyi#L186-L193) of `TypeGuard` in typeshed.

Oh, but that bit of typeshed I quoted [only just landed this morning](https://github.com/python/typeshed/pull/8057)! So we might find that this ignore is unnecessary in a future mypy release.